language/move-stdlib/nursery/sources/Capability.move

```diff
@@ -0,0 +1,153 @@
+/// A module which defines the basic concept of
+/// [*capabilities*](https://en.wikipedia.org/wiki/Capability-based_security) for managing access control.
+///
+/// # Overview
+///
+/// A capability is a unforgeable token which testifies that a signer has authorized a certain operation.
+/// The token is valid during the transaction where is is obtained. Since the type `Capability::Cap` has
+/// no ability to be stored in global memory, capabilities cannot leak out of a transaction. For every function
+/// called within a transaction which has a capability as a parameter, it is guaranteed that the capability
+/// has been obtained via a proper signer-based authorization step previously in the transaction's execution.
+///
+/// ## Basic Usage
+///
+/// Capabilities are used typically as follows:
+///
+/// ```
+///   struct ProtectedFeature { ... } // this can be just a type tag, or actually some protected data
+///
+///   public fun initialize(s: &signer) {
+///     // Create capability. This happens once at module initialization time.
+///     Capability::create<ProtectedFeature>(s);
+///   }
+///
+///   public fun do_something(s: &signer) {
+///     // Acquire the capability. This is the authorization step. Must have a signer to do so.
+///     let cap = Capability::acquire<ProtectedFeature>(s);
+///     // Pass the capability on to functions which require authorization.
+///     critical(cap);
+///   }
+///
+///   fun critical(cap: Capability::Cap<ProtectedFeature>) {
+///     // Authorization guaranteed by construction -- no verification needed!
```

Comment:
More specifically, what doesn't need be verified is: "this function is called by either the cap owner or a cap delegate whom the owner previously delegated the cap to." However, there are still something that needs to be checked. For example,
- No other function without `cap` performs the privileged operation.
- A client code uses `Capability::delegate` to an intended user (MintCap is given to a TC, not other role).

Reply:
> More specifically, what doesn't need be verified is: "this function is called by either the cap owner or a cap delegate whom the owner previously delegated the cap to." However, there are still something that needs to be checked. For example,
> 
> * No other function without `cap` performs the privileged operation (because `cap` and the privileged operation are not being tied, although Vault will do).

Good point. I addressed this partially by adding a new section "Advanced Auth Scenarios" to the documentation, describing a pattern which over module privacy would restrict who can acquire a capability. But I would agree that we should also think about ways how we can define a predicate as part of a global invariant, like the current approach you have been working on.  

> * A client code uses `Capability::delegate` to an intended user (MintCap is given to a TC, not to other role).

I think this can be addressed, see the section about delegation, and the example invariants.